tools/distpartitioning/gloo_wrapper.py

```diff
@@ -59,6 +59,74 @@ def alltoall_cpu(rank, world_size, output_tensor_list, input_tensor_list):
     for i in range(world_size):
         dist.scatter(output_tensor_list[i], input_tensor_list if i == rank else [], src=i)
 
+def alltoallv_cpu_data(rank, world_size, input_tensor_list, val_dtype):
+    """
+    Wrapper function to providing the alltoallv functionality by using underlying alltoall
+    messaging primitive
+
+    Parameters:
+    -----------
+    rank : int
+        The rank of current worker
+    world_size : int
+        The size of the entire
+    input_tensor_list : List of tensor
+        The tensors to exchange
+    val_dtype : torch dtype
+        torch dtype which is used to create buffer for receiving messages
+
+    Returns:
+    --------
+    list : 
+        list of tensors received from other processes during alltoall message
+
+    """
+    sizes = [list(x.size()) for x in input_tensor_list]
+    for idx in range(1,len(sizes)):
+        assert len(sizes[idx-1]) == len(sizes[idx])
+
+    #decide how much to pad. 
+    #always use the first-dimension for padding. 
+    ll = [ x[0] for x in sizes ]
+
+    #dims of the outgoing messages
+    out_dims = [ [np.amax(ll)] + l[1:] for idx, l in enumerate(sizes) ]
+
+    #dims of the padding needed, if any
+    diff_dims = [ [np.amax(ll) - l[0]] + l[1:] for l in sizes ]
+
+    #pad the actual message
+    input_tensor_list = [torch.cat((x, torch.zeros(diff_dims[idx]).type(x.dtype))) for idx, x in enumerate(input_tensor_list)]
+
+    #send useful message sizes to all
+    send_counts = []
+    recv_counts = []
+    for idx in range(world_size):
+        #send a 3 element triplet, [a, b, ....] where 
+        #a = useful message dim, b = amount of padding and remaining elements are
+        #the remaining dimensions of the tensor
+        send_counts.append(torch.from_numpy(np.array([sizes[idx][0]] + out_dims[idx])).type(torch.int64))
```

Comment:
how is b the amount of padding? isn't b the total size of the first dimension after padding?

Reply:
alltoall should send tensors of the same shape, right? out_dims for each rank should be the same?

Reply:
if so, you only need to exchange `sizes[idx][0]` of each rank.

Reply:
'b' is the actual message size along the first dimension. The wording in this comment were incorrect. I changed the wording to as mentioned above. 

alltoall should send tensors of same size. Yes out_dims should all be the same. 

sizes[idx][0] is the size of the input tensor along first dimension which is unpadded. out_dim[idx][0] is the size of the input tensor with padding along first dimension. 

Agree. we only need to exchange sizes[idx][0], which is just one integer as opposed to 3 integers. 

After addressing all these comments, and code modified in response to these comments. Following changes are made. 

1. Removed out_dims, as it is not needed. 
2. exchange only sizes[idx][0] which is the actual size along the first dimension (without padding).